src/java/org/apache/cassandra/index/sai/disk/io/IndexFileUtils.java

```diff
@@ -77,7 +77,7 @@ public IndexInput openBlockingInput(File file)
 
     public interface ChecksumWriter
     {
-        long getChecksum();
+        long getChecksum() throws IOException;
     }
 
     static class IncrementalChecksumSequentialWriter extends SequentialWriter implements ChecksumWriter
```

Comment:
I'm actually not sure why we have the `ChecksumWriter` interface at all here. (CC @mike-tr-adamson) Also not sure what's incremental about this now, given we just write it on flush.

There is the class `ChecksummedSequentialWriter` in `o.a.c.io.util`, and that even shares essentially the same `flushData()` method (but it writes to a concrete class called `ChecksumWriter` which has an incremental component as well as its own `DataOutput` to another file).

Perhaps we could address the `flushData()` duplication by having a `SequentialWriter` subclass w/ an abstract method that updates the checksum in a subclass-appropriate way. Then you'd have two subclasses, one strictly in memory for this case, and the other essentially doing what `o.a.c.io.util.ChecksummedSequentialWriter` does now. This might not be the only way to address this. (Naming would be interesting, as usual. `InMemoryChecksumSequentialWriter` is awful, but sort of describes what the one in this class does...)

Reply:
The idea of bypassing this whole problem by checksumming on flush is great though :D

Reply:
> Perhaps we could address the flushData() duplication by having a SequentialWriter subclass w/ an abstract method that updates the checksum in a subclass-appropriate way.

I don't agree with that. Subclassing, especially subclassing with implementation inheritance / overrides, introduces more complexity and causes way more pain than a few duplicated lines of code. In this particular case, there were several issues caused by `writeXXX` overrides which is a textbook example of inheritance gone bad. If I'm going to do a refactoring, it would be reducing the amount of inheritance, not adding more of it ;)

`ChecksummedSequentialWriter` looks like a class that has too many responsibilities. It computes the checksum *and* writes checksum to a separate component. It looks what we really want is to split those responsibilities, and really have a separate `ChecksummedSequentialWriter` that does only the checksummng part, and a separate optional component/function responsible for writing the checksum to a file. If I refactor code like that, then I could get rid of duplication, because we'd have only one checksumming writer (and no subclasses).

Reply:
Yeah, it is doing a lot of things. In any case, I like that approach better than what I’ve suggested…

Reply:
Just looked at it again and I'm afraid it is not so easy - this `ChecksummedSequentialWriter` actually writes a series of checksums to separate file. So it is kinda doing a different thing than what we want (at the moment) in SAI. SAI is just interested in computing one checksum. Note that each writer uses even a different structure to compute the checksum. I'm afraid that trying to extract the common logic between those two or trying to blend those two usecases into a single class is going to actually make the code more complex and less readable, even if slghtly less duplicated.

Reply:
I simplified `flushData` and removed the redundant one-impl interface.